dotnet/src/Microsoft.Agents.AI.AGUI/AGUIChatClient.cs

```diff
@@ -0,0 +1,325 @@
+                        // We store it on additional properties of the function call content, which will be passed down
+                        // in the next turn.
+                        fcc.AdditionalProperties ??= [];
+                        fcc.AdditionalProperties["agui_thread_id"] = threadId;
+                    }
+                    else
+                    {
+                        // Hide the server result call from the FunctionInvokingChatClient.
+                        // The wrapping client will unwrap it and present it as a normal function result.
+                        update.Contents[0] = new ServerFunctionCallContent(fcc);
+                    }
+                }
+
+                // Remove the conversation id before yielding so that the wrapping FunctionInvokingChatClient
+                // sends the whole message history on every turn as per AG-UI requirements.
+                update.ConversationId = null;
+                yield return update;
+            }
+        }
+
+        // Extract the thread id from the options additional properties
+        private static string? ExtractThreadIdFromOptions(ChatOptions? options)
+        {
+            if (options?.AdditionalProperties == null ||
+              !options.AdditionalProperties.TryGetValue("agui_thread_id", out var threadIdObj) ||
+              threadIdObj is not string threadId ||
+              string.IsNullOrEmpty(threadId))
+            {
+                return null;
+            }
+            return threadId;
+        }
+
+        // Extract the thread id from the second last message's function call content additional properties
+        private static string? ExtractTemporaryThreadId(List<ChatMessage> messagesList)
+        {
+            if (messagesList.Count < 2)
+            {
+                return null;
+            }
+            var functionCall = messagesList[messagesList.Count - 2];
+            if (functionCall.Contents.Count < 1 || functionCall.Contents[0] is not FunctionCallContent content)
+            {
+                return null;
+            }
+
+            if (content.AdditionalProperties == null ||
+              !content.AdditionalProperties.TryGetValue("agui_thread_id", out var threadIdObj) ||
+              threadIdObj is not string threadId ||
+              string.IsNullOrEmpty(threadId))
+            {
+                return null;
+            }
+
+            return threadId;
+        }
+
+        public void Dispose()
+        {
+            // No resources to dispose
+        }
+
+        public object? GetService(Type serviceType, object? serviceKey = null)
+        {
+            if (serviceType == typeof(ChatClientMetadata))
+            {
+                return this.Metadata;
+            }
+
+            return null;
+        }
+    }
+
+    private class ServerFunctionCallContent(FunctionCallContent functionCall) : AIContent
```

Comment:
If I'm understanding correctly, you're just using this as an internal implementation detail to bypass the FICC's invocation? You wrap an FCC before handing it back through the FICC and then unwrap it after it passes through the FICC and before actually returning it to the caller? Just verifying because if it were to escape to anything that might try to serialize it, we'll have to do more.